File: dunning-never-retries-a-failed-charge/python/retry_failed_billing.py
"""Retry Shopify subscription billing attempts that failed, on a safe backoff.

A recurring charge can fail for many ordinary reasons: an expired card, a bank
that declined the charge, a payment gateway timeout. Shopify records the failure
on the SubscriptionContract as lastPaymentStatus FAILED, but nothing retries it
on its own. This job finds active contracts whose last payment failed, looks at
the contract's own billingAttempts history to work out how long it has been
failing, and creates a new billing attempt once the right number of days have
passed, following a backoff schedule so we do not hammer a card that just failed.
Read the billing history, decide with a pure function, then only write
(subscriptionBillingAttemptCreate) when it is due. Run on a schedule, for
example daily. Safe to run again and again because a due contract that already
has a fresh pending attempt is left alone.
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("retry_failed_billing")

SHOP = os.environ["SHOPIFY_SHOP"]
TOKEN = os.environ["SHOPIFY_ACCESS_TOKEN"]
API_VERSION = os.environ.get("SHOPIFY_API_VERSION", "2025-01")
ENDPOINT = f"https://{SHOP}/admin/api/{API_VERSION}/graphql.json"
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

# Days to wait after the most recent attempt before trying again. The index in
# this list is the retry number: the 1st retry waits 1 day, the 2nd waits 3
# days, the 3rd waits 7 days. After that we stop retrying and leave the
# contract for a human, since a card that has failed four times in under two
# weeks needs a person, not another automatic charge.
BACKOFF_SCHEDULE_DAYS = [1, 3, 7]
MAX_RETRIES = len(BACKOFF_SCHEDULE_DAYS)

CONTRACTS_QUERY = """
query($cursor: String) {
  subscriptionContracts(first: 25, after: $cursor, query: "status:ACTIVE") {
    pageInfo { hasNextPage endCursor }
    nodes {
      id
      lastPaymentStatus
      customer { defaultEmailAddress { emailAddress } }
      billingAttempts(first: 10, reverse: true) {
        nodes { id createdAt completedAt }
      }
    }
  }
}"""

RETRY_MUTATION = """
mutation($contractId: ID!, $idempotencyKey: String!) {
  subscriptionBillingAttemptCreate(
    subscriptionContractId: $contractId
    subscriptionBillingAttemptInput: { idempotencyKey: $idempotencyKey }
  ) {
    subscriptionBillingAttempt { id }
    userErrors { field message }
  }
}"""


def gql(query, variables=None):
    r = requests.post(
        ENDPOINT,
        json={"query": query, "variables": variables or {}},
        headers={"X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    body = r.json()
    if body.get("errors"):
        raise RuntimeError(body["errors"])
    return body["data"]


def _parse_iso(value):
    if not value:
        return None
    return value.replace("Z", "+00:00")


def failed_attempt_count(billing_attempts):
    """How many attempts in a row, most recent first, never completed.

    An attempt with completedAt set was successful and produced an order, so a
    run of failures stops at the first completed attempt. billing_attempts must
    already be ordered most recent first (reverse: true on the query).
    """
    count = 0
    for attempt in billing_attempts or []:
        if attempt.get("completedAt"):
            break
        count += 1
    return count


def days_between(earlier_iso, later_iso):
    """Whole days between two ISO 8601 timestamps, later minus earlier."""
    from datetime import datetime

    earlier = datetime.fromisoformat(_parse_iso(earlier_iso))
    later = datetime.fromisoformat(_parse_iso(later_iso))
    return (later - earlier).total_seconds() / 86400.0


def retry_decision(contract, now_iso):
    """Pure decision: should we fire another billing attempt for this contract, right now?

    Rules, in order:
      1. Only contracts whose lastPaymentStatus is FAILED are candidates. A
         contract that is paid up has nothing to retry.
      2. We count the unbroken run of failed attempts from most recent
         backwards. If that count is already at or beyond MAX_RETRIES, we stop
         retrying automatically and leave it for a human.
      3. We look at how many whole days have passed since the most recent
         attempt. The schedule says how long to wait before the next retry
         (schedule[failed_count - 1], since failed_count is 1-indexed by the
         time we get here). If not enough days have passed yet, it is not due.
      4. If there is no billing attempt history at all, there is nothing to
         retry against, so we skip.

    No I/O happens in this function, so it is fully unit testable.
    """
    if contract.get("lastPaymentStatus") != "FAILED":
        return False

    attempts = contract.get("billingAttempts") or []
    failed_count = failed_attempt_count(attempts)
    if failed_count == 0 or failed_count > MAX_RETRIES:
        return False

    last_attempt = attempts[0]
    last_created_at = last_attempt.get("createdAt")
    if not last_created_at:
        return False

    wait_days = BACKOFF_SCHEDULE_DAYS[failed_count - 1]
    elapsed_days = days_between(last_created_at, now_iso)
    return elapsed_days >= wait_days


def failed_contracts():
    cursor = None
    while True:
        data = gql(CONTRACTS_QUERY, {"cursor": cursor})["subscriptionContracts"]
        for node in data["nodes"]:
            yield node
        if not data["pageInfo"]["hasNextPage"]:
            return
        cursor = data["pageInfo"]["endCursor"]


def retry_billing(contract_id, idempotency_key):
    result = gql(RETRY_MUTATION, {"contractId": contract_id, "idempotencyKey": idempotency_key})[
        "subscriptionBillingAttemptCreate"
    ]
    if result["userErrors"]:
        raise RuntimeError(result["userErrors"])
    return result["subscriptionBillingAttempt"]["id"]


def run():
    from datetime import datetime, timezone
    import uuid

    now_iso = datetime.now(timezone.utc).isoformat()
    retried = 0
    for contract in failed_contracts():
        if not retry_decision(contract, now_iso):
            continue
        contract_id = contract["id"]
        idempotency_key = f"dunning-retry-{contract_id.rsplit('/', 1)[-1]}-{uuid.uuid4().hex[:8]}"
        log.info(
            "Contract %s is due for a retry. %s",
            contract_id,
            "would retry" if DRY_RUN else "retrying",
        )
        if not DRY_RUN:
            retry_billing(contract_id, idempotency_key)
        retried += 1
    log.info("Done. %d contract(s) %s.", retried, "to retry" if DRY_RUN else "retried")


if __name__ == "__main__":
    run()

File: dunning-never-retries-a-failed-charge/node/retry-failed-billing.js
/**
 * Retry Shopify subscription billing attempts that failed, on a safe backoff.
 *
 * A recurring charge can fail for many ordinary reasons: an expired card, a bank
 * that declined the charge, a payment gateway timeout. Shopify records the failure
 * on the SubscriptionContract as lastPaymentStatus FAILED, but nothing retries it
 * on its own. This job finds active contracts whose last payment failed, looks at
 * the contract's own billingAttempts history to work out how long it has been
 * failing, and creates a new billing attempt once the right number of days have
 * passed, following a backoff schedule so we do not hammer a card that just failed.
 * Read the billing history, decide with a pure function, then only write
 * (subscriptionBillingAttemptCreate) when it is due. Run on a schedule, for
 * example daily.
 *
 * Guide: https://www.allanninal.dev/shopify/dunning-never-retries-a-failed-charge/
 */
import { pathToFileURL } from "node:url";
import { randomUUID } from "node:crypto";

const SHOP = process.env.SHOPIFY_SHOP || "example.myshopify.com";
const TOKEN = process.env.SHOPIFY_ACCESS_TOKEN || "shpat_dummy";
const API_VERSION = process.env.SHOPIFY_API_VERSION || "2025-01";
const ENDPOINT = `https://${SHOP}/admin/api/${API_VERSION}/graphql.json`;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

// Days to wait after the most recent attempt before trying again. The index in
// this array is the retry number: the 1st retry waits 1 day, the 2nd waits 3
// days, the 3rd waits 7 days. After that we stop retrying automatically.
const BACKOFF_SCHEDULE_DAYS = [1, 3, 7];
const MAX_RETRIES = BACKOFF_SCHEDULE_DAYS.length;

export function failedAttemptCount(billingAttempts) {
  let count = 0;
  for (const attempt of billingAttempts || []) {
    if (attempt.completedAt) break;
    count += 1;
  }
  return count;
}

function daysBetween(earlierIso, laterIso) {
  const earlier = new Date(earlierIso).getTime();
  const later = new Date(laterIso).getTime();
  return (later - earlier) / 86400000;
}

/**
 * Pure decision: should we fire another billing attempt for this contract, right now?
 *
 * Rules, in order:
 *   1. Only contracts whose lastPaymentStatus is FAILED are candidates.
 *   2. We count the unbroken run of failed attempts from most recent backwards
 *      (billingAttempts must already be ordered most recent first). If that
 *      count is at or beyond MAX_RETRIES, we stop retrying automatically.
 *   3. We look at how many whole days have passed since the most recent
 *      attempt, and compare that against the schedule entry for this retry
 *      number. If not enough days have passed, it is not due yet.
 *   4. If there is no billing attempt history at all, there is nothing to
 *      retry against, so we skip.
 *
 * No I/O happens in this function, so it is fully unit testable.
 */
export function retryDecision(contract, nowIso) {
  if (contract.lastPaymentStatus !== "FAILED") return false;

  const attempts = contract.billingAttempts || [];
  const failedCount = failedAttemptCount(attempts);
  if (failedCount === 0 || failedCount > MAX_RETRIES) return false;

  const lastAttempt = attempts[0];
  const lastCreatedAt = lastAttempt && lastAttempt.createdAt;
  if (!lastCreatedAt) return false;

  const waitDays = BACKOFF_SCHEDULE_DAYS[failedCount - 1];
  const elapsedDays = daysBetween(lastCreatedAt, nowIso);
  return elapsedDays >= waitDays;
}

async function gql(query, variables = {}) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Shopify ${res.status}`);
  const body = await res.json();
  if (body.errors) throw new Error(JSON.stringify(body.errors));
  return body.data;
}

const CONTRACTS_QUERY = `
query($cursor: String) {
  subscriptionContracts(first: 25, after: $cursor, query: "status:ACTIVE") {
    pageInfo { hasNextPage endCursor }
    nodes {
      id
      lastPaymentStatus
      customer { defaultEmailAddress { emailAddress } }
      billingAttempts(first: 10, reverse: true) {
        nodes { id createdAt completedAt }
      }
    }
  }
}`;

const RETRY_MUTATION = `
mutation($contractId: ID!, $idempotencyKey: String!) {
  subscriptionBillingAttemptCreate(
    subscriptionContractId: $contractId
    subscriptionBillingAttemptInput: { idempotencyKey: $idempotencyKey }
  ) {
    subscriptionBillingAttempt { id }
    userErrors { field message }
  }
}`;

async function* failedContracts() {
  let cursor = null;
  while (true) {
    const data = (await gql(CONTRACTS_QUERY, { cursor })).subscriptionContracts;
    for (const node of data.nodes) yield node;
    if (!data.pageInfo.hasNextPage) return;
    cursor = data.pageInfo.endCursor;
  }
}

async function retryBilling(contractId, idempotencyKey) {
  const result = (await gql(RETRY_MUTATION, { contractId, idempotencyKey })).subscriptionBillingAttemptCreate;
  if (result.userErrors.length) throw new Error(JSON.stringify(result.userErrors));
  return result.subscriptionBillingAttempt.id;
}

export async function run() {
  const nowIso = new Date().toISOString();
  let retried = 0;
  for await (const contract of failedContracts()) {
    if (!retryDecision(contract, nowIso)) continue;
    const idempotencyKey = `dunning-retry-${contract.id.split("/").pop()}-${randomUUID().slice(0, 8)}`;
    console.log(`Contract ${contract.id} is due for a retry. ${DRY_RUN ? "would retry" : "retrying"}`);
    if (!DRY_RUN) await retryBilling(contract.id, idempotencyKey);
    retried++;
  }
  console.log(`Done. ${retried} contract(s) ${DRY_RUN ? "to retry" : "retried"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
